llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -8308,6 +8392,154 @@ llvm::Function *SPIRVToLLVM::createLibraryEntryFunc() {
+// Currently, OpLoads will yield zero and OpStores will just omit the store, if bounds checks failed. This is up to the
+// caller. This implementation handles nested structs using arrays and 1...n-dimensional arrays. This returns if we need
+// to insert a bounds checks and pointers to the newly created basic blocks (which can be optimized away by subsequent
+// passes, if not needed).
+//
+// An example: We have the following sequence in LLVM IR:
+//
+// entry:
+// %load = load <4 x float>, <4 x float> ... localArray[runtimeIndex][runtimeIndex * 2]
+//
+// which can cause a out of bounds access.
+//
+// However, we are converting it to a sequence like this:
+//
+// entry:
+//
+// %shouldLoad1 = icmp ult i32 runtimeIndex, localArray.dim1.upperBound
+// %shouldLoad2 = icmp ult i32 runtimeIndex * 2, localArray.dim2.upperBound
+// %shouldLoad = and i1 %shouldLoad1, %shouldLoad2
+// br i1 %shouldLoad, label %loadBlock, label %finalBlock
+//
+// loadBlock:
+// %load = load <4 x float>, <4 x float> ... localArray[runtimeIndex][runtimeIndex * 2]
+// br label %finalBlock
+//
+// finalBlock:
+// %finalResult = phi ... <4 x float> [ zeroinitializer, %entry], [ %load, loadBlock ]
+// ... use %finalResult subsequently
+//
+SPIRVToLLVM::ScratchBoundsCheckInfo SPIRVToLLVM::prepareScratchBoundsCheck(SPIRVValue *toBeWrapped) {
+  constexpr ScratchBoundsCheckInfo nullValue{};
+  if (!getPipelineOptions()->enableScratchAccessBoundsChecks) {
+    return nullValue;
+  }
+
+  if (!toBeWrapped || toBeWrapped->getOpCode() != OpAccessChain) {
+    return nullValue;
+  }
+
+  SPIRVStorageClassKind pointerStorageClass = toBeWrapped->getType()->getPointerStorageClass();
+
+  // We don't want bounds checks if we are outside private or function storage class.
+  if (pointerStorageClass != StorageClassPrivate && pointerStorageClass != StorageClassFunction) {
+    return nullValue;
+  }
+
+  // Fetch the access chain behind the pointer the memory operation uses.
+  const auto accessChainBase = static_cast<SPIRVAccessChainBase *const>(toBeWrapped);
+
+  llvm::SmallVector<SPIRVValue *> indicesVec;
+
+  // Find any integer index, either for accessing the array or a struct member.
+  for (auto *const operand : accessChainBase->getIndices()) {
+    assert(operand->getType()->isTypeInt());
+    indicesVec.push_back(operand);
+  }
+
+  if (indicesVec.empty()) {
+    return nullValue;
+  }
+
+  ArrayRef<SPIRVValue *> indices = indicesVec;
+  SPIRVValue *currentIndex = nullptr;
+
+  auto consumeFirstIndex = [&currentIndex, &indices]() {
+    assert(!indices.empty());
+    currentIndex = indices[0];
+    indices = indices.drop_front();
+  };
+
+  SPIRVType *baseType = accessChainBase->getBase()->getType()->getPointerElementType();
+  consumeFirstIndex();
+
+  // This will contain our final compare result.
+  Value *finalCmpResult = nullptr;
+
+  BasicBlock *entryBlock = getBuilder()->GetInsertBlock();
+  Function *parent = entryBlock->getParent();
+
+  // Build a chain of OOB checks.
```

Comment:
Can you comment whether this processes indices forwards or backwards? Right now you don't know it until the very end of the loop when `consumeFirstIndex` is called.